### PR TITLE
Copter: fix use of WP_NAVALT_MIN 

### PR DESCRIPTION
**Pull request by rmackay9**
This PR resolves issue https://github.com/ArduPilot/ardupilot/issues/12888 which can lead to the vehicle drifting away if the wind is too strong and WP_NAVALT_MIN is set too high.

The underlying cause of the issue is:

- that the current method does not use the WPNav's get_roll() and get_pitch outputs until the vehicle passes the WP_NAVALT_MIN altitude meaning the vehicle could drift horizontally a significant distance
- WPNav will stop advancing the intermediate point if the vehicle is more than the leash length from the track

These two put together mean the vehicle is not trying to move back towards the track nor is it trying to climb to the WP_NAVALT_MIN and it drifts away.

This PR resolves the problem by constantly shifting the track horizontally so that it is where the vehicle is horizontally.  Then at the last iteration just before horizontal control is enabled we move the track to be at the vehicle's "stopping point".

It might seem that we should always just reset the track to the vehicle's stopping point but if the wind was extremely high, the vehicle could end up flying horizontally quite quickly, so quickly that it's stopping point might be longer than the leash and we would be back in the position where the WPNav controller would not advance the intermediate point.

This has been tested in SITL and below are some screen shots.

Below is a picture of the vehicle's track at the moment it starts towards the first waypoint.  The wind was set to 3m/s north so the vehicle has drifted about 30m north before horizontal navigation began.
image

Here's a before image of the PSC.TPX (target North-South position) vs PX (actual North-South position) showing the vehicle drifts 200m before it passes the WP_NAVALT_MIN altitude and retakes horizontal control (i.e. the altitude climbed very slowly)
before-image

Here's an after image showing the vehicle drifts only 28m horizontally before re-taking horizontal control at the moment it passes the WP_NAVALT_MIN altitude.
after-image

Here's an after image of the roll, pitch and altitude.  We see the roll and pitch angles are relatively smooth (i.e. no sudden twitches).  Throttle output (not shown) is also smooth during the transition
after-roll-pitch

### CODE REVIEW

**Comment by rmackay9** [comment]
@VDLJu, do you think this looks OK?  Maybe you could test that it resolves your issue?

File: ArduCopter/takeoff.cpp [hunk 1]
```diff
@@ -163,52 +167,45 @@ void Mode::auto_takeoff_run()
         wp_nav->shift_wp_origin_to_current_pos();
     }
 
-    // set motors to full range
-    motors->set_desired_spool_state(AP_Motors::DesiredSpoolState::THROTTLE_UNLIMITED);
+    // check if we are not navigating because of low altitude
+    float nav_roll = 0, nav_pitch = 0;
```
Comment (lthall) [inline ArduCopter/takeoff.cpp]: For consistency this should be 0.0f.

Comment (rmackay9) [reply ArduCopter/takeoff.cpp]: fixed, thanks!

**Comment by rmackay9** [comment]
Thanks for the review.  Updated and rebased on master.  I'll merge once it passes Travis.

**Comment by VDLJu** [comment]
Hi,

Sorry I haven't replied to this and been able to test this.
Just checked your PR idea and it looks good to me. I have not checked the actual functionality or been able to test it.

Sorry for the inconvenience.

**Comment by rmackay9** [comment]
I think I will leave this bug fix out of Copter-4.0.x.  This is a rarely used feature and it can't be cleanly backported to Copter-4.0.x so I worry that there is a higher risk in making a mistake in the backport vs the bug itself.

**Pull request merged**